### PR TITLE
Fast seek implementation

### PR DESCRIPTION
**Pull request by bjuncek**
Implementation of the fast seek; rebase of the #3015
@fmassa can you please pull it in fbcode to check if ok?

rebased on the latest master

### CODE REVIEW

**Comment by bjuncek** [comment]
For some reason it's not allowing me to push it to fbsync, I've tried twice and it always sends me back to master?

File: torchvision/io/__init__.py [hunk 1]
```diff
@@ -123,19 +123,20 @@ def __next__(self):
     def __iter__(self):
         return self
 
-    def seek(self, time_s: float):
+    def seek(self, time_s: float, keyframes_only=False):
```
Comment (prabhat00155) [inline torchvision/io/__init__.py:126]: Should we have type annotation here as well?

File: test/test_video.py [hunk 2]
```diff
@@ -244,11 +244,11 @@ def _template_read_video(video_object, s=0, e=None):
     video_frames = torch.empty(0)
     frames = []
     video_pts = []
-    for frame in itertools.takewhile(lambda x: x['pts'] <= e, video_object):
-        if frame['pts'] < s:
+    for frame in itertools.takewhile(lambda x: x["pts"] <= e, video_object):
```
Comment (prabhat00155) [inline test/test_video.py:247]: Why these `' -> "` changes?

**Review by prabhat00155 — COMMENTED** [review]
Thanks @bjuncek, could you also resolve the merge conflicts?

File: test/test_video.py (hunk 2 above)
Comment (NicolasHug) [reply test/test_video.py:247]: Probably black formatting, let's please avoid these changes until we can fully use black in torchvision (Philip and I are working on it https://github.com/pytorch/vision/issues/4174)

Comment (bjuncek) [reply test/test_video.py:247]: A combination of black (or and my editor not picking up flake); sry, will revert

File: torchvision/io/__init__.py (hunk 1 above)
Comment (bjuncek) [reply torchvision/io/__init__.py:126]: yup

File: test/test_video.py [hunk 3]
```diff
@@ -402,6 +402,45 @@ def test_video_reading_fn(self):
...
+                    seek_val = (av_keyframes[i] + av_keyframes[i - 1]) / 2
+                    data = next(video_reader.seek(seek_val, True))
+                    vr_keyframes.append(data["pts"])
+
+                data = next(video_reader.seek(config.duration, True))
+                vr_keyframes.append(data["pts"])
+
+                self.assertTrue(len(av_keyframes) == len(vr_keyframes))
+                # NOTE: this video gets different keyframe with different
+                # loaders (0.333 pyav, 0.666 for us)
```
Comment (prabhat00155) [inline test/test_video.py:437]: Do we know the reason for this?

File: test/test_video.py [hunk 4]
```diff
@@ -402,6 +402,45 @@ def test_video_reading_fn(self):
...
+            av_reader = av.open(full_path)
+            # reduce streams to only keyframes
+            av_stream = av_reader.streams.video[0]
+            av_stream.codec_context.skip_frame = "NONKEY"
+
+            av_keyframes = []
+            vr_keyframes = []
+            if av_reader.streams.video:
+
+                # get all keyframes using pyav. Then, seek randomly into video reader
```
Comment (prabhat00155) [inline test/test_video.py:419]: This comment should go before the `if` statement as it describes the entire chunk of code from line 417 onwards, not just the `if` block.

**Comment by prabhat00155** [comment]
I can't import this PR internally to run tests due to merge conflicts. Could you please fix them? Thanks!

**Comment by datumbox** [comment]
@bjuncek Just wanted to clarify why the PR contains `[fbsync] ` in the title?

As far as I understand, please correct me if I'm wrong, this is a brand new PR you wrote (it doesn't come from the `fbsync` branch) which you want to have it tested on FBcode. If this is indeed the case, you need to remove the prefix because this will cause the exact opposite result of what you want. PRs tagged with `[fbsync]` are skipped from being synced to FBcode. This is because we reserve the tag for those PRs that sync FBCode=>Github.

File: torchvision/io/__init__.py [hunk 5]
```diff
@@ -130,19 +130,20 @@ def __next__(self) -> Dict[str, Any]:
     def __iter__(self) -> Iterator['VideoReader']:
         return self
 
-    def seek(self, time_s: float) -> 'VideoReader':
+    def seek(self, time_s: float, keyframes_only: bool = False):
```
Comment (prabhat00155) [inline torchvision/io/__init__.py:133]: Is there a reason you have removed annotation for return type?

**Review by prabhat00155 — APPROVED** [review]
Looks good to me, thanks @bjuncek! One minor comment.

**Comment by bjuncek** [comment]
@datumbox sorry, my bad - Francisco told me that adding fbsync tag would import it into fbcode; removed now.

**Pull request merged**